### DISCUSSION TITLE
How to implement OAuth login in React App

### DISCUSSION THREAD

**Discussion by Industrial** [directus/directus#5565] [Archive]
Hi!

I see that there are OAuth options for loggin in with Directus. I have configured Directus with env vars as described in https://docs.directus.io/reference/environment-variables/#oauth. What I would like to know if it's possible to use OAuth for getting a JWT to use in a client application / browser app instead of the admin interface.

This is what I used with [NextAuth](https://next-auth.js.org) and [STRApi](https://strapi.io/) (I am trying to move away from StrAPI and towards Directus). As you can see, the React App gets the google credentials and via an API route configured in the code below it communicates to the OAuth provider. At some point an API in the CMS is called to verify (3 legged oauth?) and a JWT is returned from this, to be used in the application/browser.
```ts
import NextAuth, { Account, Session, User } from 'next-auth'
import { JWT } from 'next-auth/jwt'
import Providers from 'next-auth/providers'

import { getAPIURL } from '../../../lib/environment'

export default NextAuth({
  debug: true,

  providers: [
    Providers.Google({
      clientId: String(process.env.GOOGLE_OAUTH_KEY),
      clientSecret: String(process.env.GOOGLE_OAUTH_SECRET),
    }),
  ],

  session: {
    jwt: true,
  },

  callbacks: {
    session: async (session: Session, user: User) => {
      return {
        ...session,

        // Exists in reality but not on the type.
        // @ts-ignore
        jwt: user.jwt,

        // Exists in reality but not on the type.
        // @ts-ignore
        userId: user.id,
      }
    },

    jwt: async (token: JWT, user?: User, account?: Account) => {
      if (!user || !account) {
        return token
      }

      const isSignedIn = Boolean(user)

      if (isSignedIn) {
        const url = `${getAPIURL()}/auth/${account.provider}/callback?access_token=${account.accessToken}`
        const response = await fetch(url)
        const data = await response.json()

        const output = {
          ...token,
          jwt: data.jwt,
          id: data.user.id,
        }

        return output
      }

      return token
    },
  },
})
```

How do I implement this with Directus?

**Comment by Industrial** [comment]
I have found https://docs.directus.io/reference/api/system/authentication/#login-using-oauth-provider but I'm not sure how to use it at the moment and I am getting the error

```
Error 400: invalid_request
Missing required parameter: client_id
```

Did I misconfigure something?

**Reply by rijkvanzanten** [reply]
Hard to tell without knowing your configuration 🙂 oAuth providers can be a little finicky in the exact data IO they expect. What provider are you testing with, and what is the configuration you're using (please redact the actual keys)

**Reply by Industrial** [reply]
Here's my package.json and .env file: https://gist.github.com/Industrial/7a6c5a1b94a2e711c27d75d1d85be4dd

**Reply by rijkvanzanten** [reply]
@Industrial Seems like you've accidentally reversed the order of the keys:

```
OAUTH_PROVIDERS="google"
OAUTH_GOOGLE_SCOPE="openid email"
>>> GOOGLE_OAUTH_KEY=KEE
>>> GOOGLE_OAUTH_SECRET=SEEKRIT
```

should be 

```
OAUTH_PROVIDERS="google"
OAUTH_GOOGLE_SCOPE="openid email"
OAUTH_GOOGLE_KEY=KEE
OAUTH_GOOGLE_SECRET=SEEKRIT
```

**Reply by Industrial** [reply]
oh, lol, right. Thanks!

**Comment by batuhanbilginn** [comment]
Did you fix this issue? I'm facing the same problem at the moment but I couldn't make it... 
`${getAPIURL()}/auth/${account.provider}/callback?access_token=${account.accessToken}`
this returns this:
`{
    "errors": [
        {
            "message": "Invalid user credentials.",
            "extensions": {
                "code": "INVALID_CREDENTIALS"
            }
        }
    ]
}`
